File: src/longbet/_iv.py
"""Convenience interface for experimental prediction-assisted LongBet IV."""
from __future__ import annotations

from typing import Any

import numpy as np

from longbet._encourage import _validate
from longbet._iv_nuisance import LongBetIVNuisance
from longbet._orthogonal_iv import CrossfitEncouragementResult, crossfit_encouragement


def baseline_features(y: Any, d: Any, z: Any, x: Any = None,
                      t: Any = None) -> np.ndarray:
    """Join supplied baseline covariates and all pre-assignment Y/D columns.

    No post-assignment response enters the feature matrix. A vector assignment
    treats all supplied periods as post-assignment, matching the inference API.
    The user must ensure that the supplied ``x`` was measured before assignment.
    No response-dependent standardization or feature selection is done here.
    """
    outcomes = np.asarray(y)
    if (outcomes.ndim != 2 or outcomes.dtype.kind not in "biuf"
            or not np.isfinite(outcomes).all()):
        raise ValueError("y must be a complete finite numeric unit-by-period matrix.")
    assignment = np.asarray(z)
    if assignment.ndim == 1:
        if len(assignment) != len(outcomes):
            raise ValueError("A vector z must contain one assignment per unit.")
        assignment = np.broadcast_to(assignment[:, None], outcomes.shape)
    panel = _validate(assignment, d, t)
    if outcomes.shape != panel.z.shape:
        raise ValueError("y, d and panel z must have identical shapes.")
    covariates = np.empty((len(outcomes), 0)) if x is None else np.asarray(x)
    if (covariates.ndim != 2 or covariates.shape[0] != len(outcomes)
            or covariates.dtype.kind not in "biuf" or not np.isfinite(covariates).all()):
        raise ValueError("x must be a complete finite numeric unit-by-covariate matrix.")
    features = np.column_stack([covariates, outcomes[:, :panel.start],
                                panel.d[:, :panel.start]]).astype(float)
    # A constant feature supports the no-baseline/no-covariate case without
    # inventing account-specific information for the prediction model.
    return features if features.shape[1] else np.ones((len(outcomes), 1))


def longbet_iv(y: Any, d: Any, z: Any, x: Any = None, t: Any = None, *,
               learner_config: dict[str, Any] | None = None,
               folds: int = 2, seed: int = 0, alpha: float = .05,
               design: str = "complete_randomization") -> CrossfitEncouragementResult:
    """Estimate assignment ITTs and complete IV confidence sets with LongBet.

    This experimental interface uses LongBet only for held-out prediction.
    ``learner_config`` contains :class:`LongBetIVNuisanceConfig` options. The
    sampling budget controls numerical prediction; posterior quantiles are not
    used for causal inference. All pre-assignment Y/D columns are appended to x.

    Supports complete individual-unit randomization and two conditional folds.
    The joint residual Neyman covariance and Fieller intervals are asymptotic
    and pointwise, requiring stable predictions and suitable moment conditions.
    A CACE interpretation additionally requires exclusion, monotonicity and
    relevance for the specified adoption history. Full unbounded/disconnected
    sets are retained. See :func:`crossfit_encouragement` for the generic learner
    interface and the design/inference assumptions.
    """
    if (isinstance(seed, (bool, np.bool_))
            or not isinstance(seed, (int, np.integer)) or seed < 0):
        raise ValueError("seed must be a nonnegative integer.")
    if learner_config is not None and not isinstance(learner_config, dict):
        raise ValueError("learner_config must be a dictionary or None.")
    options = dict(learner_config or {})
    options.setdefault("seed", int(seed) + 7919)
    # Validate options before beginning any fold fits.
    LongBetIVNuisance(**options)
    features = baseline_features(y, d, z, x, t)
    result = crossfit_encouragement(
        y, d, z, features, t, learner_factory=lambda: LongBetIVNuisance(**options),
        folds=folds, seed=seed, alpha=alpha, design=design,
    )
    result.metadata.update(
        engine="longbet_crossfit_iv", experimental=True,
        feature_source="supplied_baseline_x_plus_all_pre_assignment_y_d",
        feature_count=int(features.shape[1]), learner_seed=int(options["seed"]),
        posterior_causal_intervals=False,
    )
    return result

File: R/iv.R
#' Prediction-assisted longitudinal instrumental variables
#'
#' Experimental two-fold conditional cross-fitting with a LongBet trajectory
#' predictor. Uncertainty comes from the joint residual Neyman covariance and
#' analytic Fieller inversion, not posterior quantiles.
#'
#' @param y Complete finite outcome panel, [N x T].
#' @param d Complete binary absorbing adoption panel, [N x T].
#' @param z Binary single-wave encouragement panel, [N x T], or a vector of
#'   randomized assignments. A vector treats every supplied period as post-wave.
#' @param x Optional baseline covariates, [N x P]. All pre-wave y/d columns are
#'   appended automatically; held-out post-wave data are never used as features.
#' @param t Optional increasing observation times.
#' @param learner_config Named list of LongBetIVNuisanceConfig options. Integer
#'   sampling/tree controls must be whole numbers. The predictor selects temporal
#'   pooling on an inner account holdout within each training fold.
#' @param folds Number of folds; this prototype supports exactly two.
#' @param seed Nonnegative integer controlling the conditional split and, unless
#'   overridden in learner_config, the prediction RNG.
#' @param alpha Significance level; default 0.05.
#' @param design Must be "complete_randomization" for individual panel units.
#' @return A serializable list containing table, estimates, covariance,
#'   predictions, unit_scores, fold_ids and inference metadata. Covariance order
#'   is Y at horizon 1, D at horizon 1, Y at horizon 2, D at horizon 2, etc.
#' @details ITTs are unbiased under the stated assignment and splitting design.
#'   Covariance and intervals are asymptotic and pointwise; they require stable
#'   predictions and suitable moment conditions. Full unbounded/disconnected
#'   confidence sets are retained in lower/upper and lower2/upper2. A CACE
#'   interpretation additionally requires exclusion, monotonicity, relevance,
#'   and appropriate adoption-history restrictions. This does not implement an
#'   exact sharp-null permutation test or prove superiority over other learners.
#' @export
longbet_iv <- function(y, d, z, x = NULL, t = NULL,
                       learner_config = NULL, folds = 2L, seed = 0L,
                       alpha = 0.05, design = "complete_randomization") {
  integer_option <- function(value, name) {
    if (!is.numeric(value) || length(value) != 1L || is.na(value) ||
        !is.finite(value) || value != floor(value) ||
        value < 0 || value > .Machine$integer.max) {
      stop(name, " must be a nonnegative integer.", call. = FALSE)
    }
    as.integer(value)
  }
  folds <- integer_option(folds, "folds")
  seed <- integer_option(seed, "seed")
  if (!is.null(learner_config)) {
    if (!is.list(learner_config) ||
        (length(learner_config) && (is.null(names(learner_config)) ||
          any(!nzchar(names(learner_config))) || anyDuplicated(names(learner_config))))) {
      stop("learner_config must be a named list or NULL.", call. = FALSE)
    }
    integer_names <- c("baseline_trees", "effect_trees", "cutpoints", "burnin",
                       "draws", "chains", "max_interaction_rules", "seed")
    for (name in intersect(names(learner_config), integer_names)) {
      learner_config[[name]] <- integer_option(learner_config[[name]], name)
    }
  }
  lb <- longbet_py()
  res <- lb$longbet_iv(
    y = reticulate::r_to_py(as.matrix(y)),
    d = reticulate::r_to_py(as.matrix(d)),
    z = reticulate::r_to_py(if (is.null(dim(z))) as.numeric(z) else as.matrix(z)),
    x = if (is.null(x)) NULL else reticulate::r_to_py(as.matrix(x)),
    t = if (is.null(t)) NULL else reticulate::r_to_py(as.numeric(t)),
    learner_config = if (is.null(learner_config) || !length(learner_config)) NULL
      else reticulate::r_to_py(learner_config),
    folds = folds, seed = seed, alpha = alpha, design = design
  )
  rows <- .encouragement_as_r(res$table)
  table <- do.call(rbind, lapply(rows, as.data.frame, stringsAsFactors = FALSE))
  rownames(table) <- NULL
  list(table = table,
       estimates = .encouragement_as_r(res$estimates),
       covariance = .encouragement_as_r(res$covariance),
       predictions = .encouragement_as_r(res$predictions),
       unit_scores = .encouragement_as_r(res$unit_scores),
       fold_ids = as.integer(.encouragement_as_r(res$fold_ids)),
       metadata = .encouragement_as_r(res$metadata))
}
